cluster/engine.go

```diff
@@ -739,31 +727,69 @@ func (e *Engine) refreshVolume(IDOrName string) error {
 // true, each container will be inspected.
 // FIXME: unexport this method after mesos scheduler stops using it directly
 func (e *Engine) RefreshContainers(full bool) error {
-	e.Lock()
-	defer e.Unlock()
 	opts := types.ContainerListOptions{
 		All:  true,
 		Size: false,
 	}
+	// Get the current list of containers and cache it
+	e.RLock()
+	currentContainerIDs := make([]string, 0, len(e.containers))
+	for k := range e.containers {
+		currentContainerIDs = append(currentContainerIDs, k)
+	}
+	e.RUnlock()
+
 	ctx, cancel := context.WithTimeout(context.TODO(), requestTimeout)
 	defer cancel()
 	containers, err := e.apiClient.ContainerList(ctx, opts)
-	e.checkConnectionErr(err, true)
+	e.CheckConnectionErr(err)
 	if err != nil {
 		return err
 	}
 
+	var missingContainerIDs []string
+	containersMap := make(map[string]struct{}, len(containers))
+	for _, container := range containers {
+		containersMap[container.ID] = struct{}{}
+	}
+	for _, currentContainerID := range currentContainerIDs {
+		if _, ok := containersMap[currentContainerID]; !ok {
+			missingContainerIDs = append(missingContainerIDs, currentContainerID)
+		}
+	}
+
 	merged := make(map[string]*Container)
 	for _, c := range containers {
-		mergedUpdate, err := e.updateContainer(c, merged, full, true)
+		mergedUpdate, err := e.updateContainer(c, merged, full)
 		if err != nil {
-			log.WithFields(log.Fields{"name": e.Name, "id": e.ID}).Errorf("Unable to update state of container %q: %v", c.ID, err)
+			if strings.Contains(err.Error(), "No such container") {
+				// The container might have been deleted after ContainerList was returned and before reaching here
+				// Confirm it's benign by checking the e.containers
+				e.RLock()
+				for containerID := range e.containers {
```

Comment:
`e.containers` is a map, right? So can't you just do `_, ok := e.containers[c.ID]` to check if `c.ID` is in `e.containers`?

Reply:
yep :( another oversight, I just pushed an updated.
I will run a test with this changed and confirm everything is ok